connector/proto/src/main/scala/org/apache/spark/sql/proto/CatalystDataToProto.scala

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.proto
+
+import com.google.protobuf.DynamicMessage
+
+import org.apache.spark.sql.catalyst.expressions.{Expression, UnaryExpression}
+import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, ExprCode}
+import org.apache.spark.sql.proto.utils.{ProtoUtils, SchemaConverters}
+import org.apache.spark.sql.types.{BinaryType, DataType}
+
+private[proto] case class CatalystDataToProto(
+                                               child: Expression,
+                                               descFilePath: Option[String],
+                                               messageName: Option[String])
+  extends UnaryExpression {
+
+  override def dataType: DataType = BinaryType
+
+  @transient private lazy val protoType = (descFilePath, messageName) match {
+      case (Some(a), Some(b)) => ProtoUtils.buildDescriptor(a, b)
+      case _ => SchemaConverters.toProtoType(child.dataType, child.nullable)
```

Comment:
What does it mean to serialize without a proto definition. How can we create a protobuf? How can consumers read it? Looks like it will be hard to ensure backward or forward compatibility (which is fundamental to protobufs).

Reply:
We support both with and without a descriptor. I agree that deserializing (from_protobuf) without a protobuf descriptor will not work. It was mainly added to maintain symmetry with to_avro which supports without a schema.

Reply:
I would say lets not support it. Users will use it without realizing the limitations and will be confused. Protobufs are never meant be used without access to schema.

Reply:
The actual arguments for these functions would be a bit different compared to Avro functions. It is ok to have the differences, we just need to document the functions well.

Reply:
Okay, we can remove this support in to_avro. A schema will always be required.